redmine_import_time_entry/unit/import_synchronizer.py

```diff
@@ -0,0 +1,122 @@
+from openerp.tools.translate import _
+
+from openerp.addons.connector.exception import ConnectorException
+
+from openerp.addons.connector_redmine.backend import redmine
+from openerp.addons.connector_redmine.connector import get_environment
+from openerp.addons.connector_redmine.unit.import_synchronizer import (
+    RedmineBatchImportSynchronizer, RedmineImportSynchronizer,
+    import_record)
+
+from datetime import datetime
+from tools import ustr
+
+
+@redmine
+class TimeEntryBatchImportSynchronizer(RedmineBatchImportSynchronizer):
+
+    _model_name = 'redmine.hr.analytic.timesheet'
+
+    def run(self, filters=None, options=None):
+        """ Run the synchronization """
+        if options is None:
+            options = {}
+
+        updated_from = datetime.strptime(
+            self.backend_record.time_entry_last_update,
+            DEFAULT_SERVER_DATETIME_FORMAT)
+
+        if options.get('single_user', False):
+            func = import_record
+
+            login = filters.pop('login')
+            user_id = self.backend_adapter.search_user(login)
+            filters['user_id'] = user_id
+
+        else:
+            func = import_record.delay
+
+        record_ids = self.backend_adapter.search(
+            updated_from, filters)
+
+        session = self.session
+        model_name = self._model_name
+        backend_id = self.backend_record.id
+
+        for record_id in record_ids:
+            func(session, model_name, backend_id, record_id, options=options)
+
+
+@redmine
+class TimeEntryImportSynchronizer(RedmineImportSynchronizer):
+
+    _model_name = 'redmine.hr.analytic.timesheet'
+
+    def run(self, record_id, options=None):
+        """
+        Update the last synchronization date on the backend record
+
+        The schedule date of the import batch in Odoo can not be
+        used because the time zone can be different in both systems.
+
+        Each time a batch is imported, we take the record with the
+        highest value for updated_on.
+        """
+        if options is None:
+            options = {}
+
+        super(TimeEntryImportSynchronizer, self).run(record_id)
+
+        backend = self.backend_record
+
+        if (
+            self.updated_on > backend.time_entry_last_update and
+            not options.get('single_user', False)
+        ):
+            backend.write({'time_entry_last_update': self.updated_on})
+
+
+def import_single_user_time_entries(
+    session, backend_id, login, date_from, date_to
+):
+    """ Import time entries for a single user """
+    env = get_environment(session, 'redmine.hr.analytic.timesheet', backend_id)
+    importer = env.get_connector_unit(RedmineBatchImportSynchronizer)
+
+    filters = {
+        'login': login,
+        'from_date': date_from,
+        'to_date': date_to,
+    }
+
+    try:
+        importer.run(filters=filters, options={'single_user': True})
+    except ConnectorException as err:
+        raise orm.except_orm(
+            _('Error !'),
```

Comment:
In english, there are no spaces before punctuation. In fact, an exclamation mark is a little overkill for a title.